### PR TITLE
Fix firstEventBlockNumber bug in metadata

### PR DESCRIPTION
**Pull request by JonoPrest**
There was a bug introduced with partitioned fetchState where the firstEventBlockNumber can be set to any partitions first block.

### CODE REVIEW

File: codegenerator/cli/templates/static/codegen/src/globalState/GlobalState.res [hunk 1]
```diff
@@ -307,11 +307,6 @@ let handleBlockRangeResponse = (state, ~chain, ~response: ChainWorker.blockRange
       ->Utils.unwrapResultExn
       ->updateChainFetcherCurrentBlockHeight(~currentBlockHeight)
 
-    let firstEventBlockNumber = switch parsedQueueItems[0] {
-    | Some(item) if chainFetcher.firstEventBlockNumber->Option.isNone => item.blockNumber->Some
-    | _ => chainFetcher.firstEventBlockNumber
-    }
-
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/globalState/GlobalState.res:314]: Previously this number was just being set on chain fetcher here. Either it existed or it didn't. Now that there can be multiple partitions even when the indexer starts up, the first batch to come back would not necessarily contain the first event.

File: codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res [hunk 2]
```diff
@@ -95,6 +95,7 @@ type rec t = {
   //of a rollback.
   dynamicContracts: DynamicContractsMap.t,
   isFetchingAtHead: bool,
+  firstEventBlockNumber: option<int>,
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res:98]: Solution is to add firstEventBlockNumber to each partition and evaluate the earliest across partitions.

File: codegenerator/cli/templates/static/codegen/src/eventFetching/PartitionedFetchState.res [hunk 3]
```diff
@@ -273,3 +273,9 @@ let isFetchingAtHead = ({partitions}: t) => {
     accum && partition.isFetchingAtHead
   })
 }
+
+let getFirstEventBlockNumber = ({partitions}: t) => {
+  partitions->List.reduce(None, (accum, partition) => {
+    Utils.Math.minOptInt(accum, partition.firstEventBlockNumber)
+  })
+}
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/PartitionedFetchState.res:281]: Evaluate each partition's firstEventBlockNumber

File: codegenerator/cli/templates/static/codegen/src/eventFetching/ChainFetcher.res [hunk 4]
```diff
@@ -375,3 +375,8 @@ let rollbackToLastBlockHashes = (chainFetcher: t, ~rolledBackLastBlockData) => {
 let isFetchingAtHead = (chainFetcher: t) =>
   chainFetcher.fetchState->PartitionedFetchState.isFetchingAtHead
 
+let getFirstEventBlockNumber = (chainFetcher: t) =>
+  Utils.Math.minOptInt(
+    chainFetcher.dbFirstEventBlockNumber,
+    chainFetcher.fetchState->PartitionedFetchState.getFirstEventBlockNumber,
+  )
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/ChainFetcher.res:382]: Evaluate the DBs first event block number against the partions first event block number in the case of an indexer restart.

File: codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res [hunk 5]
```diff
@@ -199,11 +204,23 @@ exception UnexpectedRegisterDoesNotExist(id)
...
-  fetchedEventQueue: List.concat(self.fetchedEventQueue, newFetchedEvents),
+let updateRegister = (
+  self: t,
+  ~latestFetchedBlock,
+  ~newFetchedEvents: list<Types.eventBatchQueueItem>,
+  ~isFetchingAtHead,
+) => {
+  let firstEventBlockNumber = switch self.firstEventBlockNumber {
+  | Some(n) => Some(n)
+  | None => newFetchedEvents->Utils.List.getLast->Option.map(v => v.blockNumber)
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res:215]: Why do you get last here instead of the first?

**Review by DZakh — APPROVED** [review]
Besides the `newFetchedEvents->Utils.List.getLast` line everything looks good 👍

And I don't like the `list` usage, but it's not related to the PR

File: codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res (hunk 5 above)
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res:215]: Ah good catch sorry I had it backwards in my head!

File: codegenerator/cli/templates/static/codegen/src/Utils.res [hunk 6]
```diff
@@ -88,6 +88,23 @@ module Dict = {
    */
   external dangerouslyGetNonOption: (dict<'a>, string) => option<'a> = ""
 }
+module List = {
+  let rec getLast = (list, ~head=?) =>
+    switch list {
+    | list{} => head
+    | list{head, ...tail} => tail->getLast(~head)
+    }
```
[the remark this answers is not in the document]
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/Utils.res:96]: Hey @DZakh, I agree array is generally better. Lets have a chat, some time there's trade off in both. Array concats are quite heavy for immutable operation and other types of traversal that currently happen.

Some places can probably moved back to array.

**Comment by JonoPrest** [comment]
@DZakh I've updated the code for getting the list item. I had it backwards in my head (was writing for an ordering of latest to earliest event not earliest first which is the correct)

**Comment by DZakh** [comment]
Looks good 👌

**Pull request merged**